### PR TITLE
issue #5043, issue #27331, issue #19111, issue #28159 - Characteristics minor refactor

### PR DESCRIPTION
**Pull request by andersdd**
resolves issues #5043, #27331, #19111, #28159

### CODE REVIEW

File: foundation-database/public/functions/createinvoice.sql [hunk 1]
```diff
@@ -130,6 +130,16 @@ BEGIN
...
 
+--  Create the Invoice Item Characteristic Assignments
+    INSERT INTO charass
+          (charass_target_type, charass_target_id, charass_char_id, charass_value, charass_default, charass_price)
+    SELECT 'INVI', _invcitemid, charass_char_id, charass_value, charass_default, charass_price
+    FROM coitem
+      JOIN charass ON charass_target_type = 'SI' AND charass_target_id = coitem_id
+      JOIN char    ON char_id = charass_char_id
+      JOIN charuse ON char_id = charuse_char_id AND charuse_target_type = 'INVI'
+    WHERE coitem_id = _r.coitem_id;
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/createinvoice.sql]: `char` isn't necessary here as it's just as easy to
```sql
join charuse on charass_char_id = charuse_char_id and charuse_target_type = 'INVI'
```

File: foundation-database/public/trigger_functions/charass.sql [hunk 2]
```diff
@@ -97,13 +97,9 @@ CREATE TRIGGER charassHistoryTrigger BEFORE INSERT OR UPDATE OR DELETE ON charas
...
 CREATE OR REPLACE FUNCTION _charassuniquetrigger()
   RETURNS trigger AS $$
--- Copyright (c) 1999-2016 by OpenMFG LLC, d/b/a xTuple.
+-- Copyright (c) 1999-2017 by OpenMFG LLC, d/b/a xTuple.
 -- See www.xtuple.com/CPAL for the full text of the software license.
 BEGIN
-  IF (NEW.charass_target_type = 'LS') THEN
-  -- Exclusion for Lot/Serial characteristics as they are retained in the system (#27761)
-    RETURN NEW;
-  END IF;
```
Comment (gilmoskowitz) [inline foundation-database/public/trigger_functions/charass.sql]: What about issue #27761?

Comment (andersdd) [inline foundation-database/public/trigger_functions/charass.sql]: I believe the recent inventory refactor has removed the need to exclude LS type from this logic.  I tested various lot scenarios and we are no longer duplicating the serial # in tables and thus not duplicating the characteristics against that lot.  I could not get this to fail in testing.

Comment (andersdd) [inline foundation-database/public/trigger_functions/charass.sql]: Retested #27761 with revised code and no errors encountered - although possibly uncovered another issue.

**Pull request merged**